Special character encoding error in YAML file not properly identified

**Home Assistant release (`hass --version`):**
0.30.2

**Python release (`python3 --version`):**
3.4.2

**Description of problem:**
Was trying to create an average main floor temperature sensor using a value template.  Found this example here:  https://community.home-assistant.io/t/get-average-of-multiple-temperature-sensors/4050/2?u=rpitera

I copied/pasted the example into my sensors.yaml file after adjusting the sensors to reflect my own and changing the unit of measurement from C to F.   Upon restarting, HA was unable to start and would not produce an error log.  Running `sudo systemctl status home-assistant.service -l` produced this error:

```
● home-assistant.service - Home Assistant
Loaded: loaded (/etc/systemd/system/home-assistant.service; enabled)
Active: failed (Result: exit-code) since Mon 2016-10-17 23:52:26 EDT; 20s ago
Process: 2278 ExecStart=/srv/hass/hass_venv/bin/hass -c /home/hass/ (code=exited, status=1/FAILURE)
Main PID: 2278 (code=exited, status=1/FAILURE)
Oct 17 23:52:25 raspi1 hass[2278]: self.update(index+1)
Oct 17 23:52:25 raspi1 hass[2278]: File "/srv/hass/hass_venv/lib/python3.4/site-packages/yaml/reader.py", line 153, in update
Oct 17 23:52:25 raspi1 hass[2278]: self.update_raw()
Oct 17 23:52:25 raspi1 hass[2278]: File "/srv/hass/hass_venv/lib/python3.4/site-packages/yaml/reader.py", line 178, in update_raw
Oct 17 23:52:25 raspi1 hass[2278]: data = self.stream.read(size)
Oct 17 23:52:25 raspi1 hass[2278]: File "/srv/hass/hass_venv/lib/python3.4/codecs.py", line 313, in decode
Oct 17 23:52:25 raspi1 hass[2278]: (result, consumed) = self._buffer_decode(data, self.errors, final)
Oct 17 23:52:25 raspi1 hass[2278]: UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb0 in position 179: invalid start byte
Oct 17 23:52:26 raspi1 systemd1: home-assistant.service: main process exited, code=exited, status=1/FAILURE
Oct 17 23:52:26 raspi1 systemd1: Unit home-assistant.service entered failed state.
```
With help from a forum member I figured out that the issue was the degree symbol that I copied in the example file from the forum post.  But the file that it was in - sensors.yaml - is not referenced in the error above.  Without guidance in the forum mentioning to look for special characters, I would never have known to look at the degree symbol as being the culprit.

**Expected:**

HA should be able to trap formatting issue like this and produce an error in an error log rather than just not starting up and giving no indication where the issue is.  It should at least identify the file that has the issue.  Using the config check script method showed nothing either.

In my initial trouble shooting steps, I immediately thought something in the new sensor was causing the issue (since it was the only new change) so I commented it out of the sensors.yaml file, but HA still refused to start.  Not sure why HA would read the section if it was commented out, so this seems like an issue as well.

```
pi@raspi1:~$ pip install git+https://github.com/home-assistant/home-assistant.git@13ab2be5f6e81c697e437b425841368e32ce1948
Downloading/unpacking git+https://github.com/home-assistant/home-assistant.git@13ab2be5f6e81c697e437b425841368e32ce1948
  Cloning https://github.com/home-assistant/home-assistant.git (to 13ab2be5f6e81c697e437b425841368e32ce1948) to /tmp/pip-tYIAq0-build
  Could not find a tag or branch '13ab2be5f6e81c697e437b425841368e32ce1948', assuming commit.
  Running setup.py (path:/tmp/pip-tYIAq0-build/setup.py) egg_info for package from git+https://github.com/home-assistant/home-assistant.git@13ab2be5f6e81c697e437b425841368e32ce1948

Requirement already satisfied (use --upgrade to upgrade): requests<3,>=2 in /usr/local/lib/python2.7/dist-packages (from homeassistant==0.32.0.dev0)
Downloading/unpacking pyyaml<4,>=3.11 (from homeassistant==0.32.0.dev0)
  Downloading PyYAML-3.12.tar.gz (253kB): 253kB downloaded
  Running setup.py (path:/tmp/pip-build-7bLi6E/pyyaml/setup.py) egg_info for package pyyaml

Downloading/unpacking pytz>=2016.6.1 (from homeassistant==0.32.0.dev0)
  Downloading pytz-2016.7-py2.py3-none-any.whl (480kB): 480kB downloaded
Downloading/unpacking pip>=7.0.0 (from homeassistant==0.32.0.dev0)
  Downloading pip-8.1.2-py2.py3-none-any.whl (1.2MB): 1.2MB downloaded
Downloading/unpacking jinja2>=2.8 (from homeassistant==0.32.0.dev0)
  Downloading Jinja2-2.8-py2.py3-none-any.whl (263kB): 263kB downloaded
Downloading/unpacking voluptuous==0.9.2 (from homeassistant==0.32.0.dev0)
  Downloading voluptuous-0.9.2.tar.gz
  Running setup.py (path:/tmp/pip-build-7bLi6E/voluptuous/setup.py) egg_info for package voluptuous
    WARNING: Could not locate pandoc, using Markdown long_description.

Downloading/unpacking typing<4,>=3 (from homeassistant==0.32.0.dev0)
  Downloading typing-3.5.2.2.tar.gz (51kB): 51kB downloaded
  Running setup.py (path:/tmp/pip-build-7bLi6E/typing/setup.py) egg_info for package typing

Downloading/unpacking MarkupSafe (from jinja2>=2.8->homeassistant==0.32.0.dev0)
  Downloading MarkupSafe-0.23.tar.gz
  Running setup.py (path:/tmp/pip-build-7bLi6E/MarkupSafe/setup.py) egg_info for package MarkupSafe

Requirement already satisfied (use --upgrade to upgrade): setuptools>=0.6b1 in /usr/local/lib/python2.7/dist-packages (from voluptuous==0.9.2->homeassistant==0.32.0.dev0)
Installing collected packages: pyyaml, pytz, pip, jinja2, voluptuous, typing, homeassistant, MarkupSafe
  Running setup.py install for pyyaml
    checking if libyaml is compilable
    arm-linux-gnueabihf-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I/usr/include/python2.7 -c build/temp.linux-armv7l-2.7/check_libyaml.c -o build/temp.linux-armv7l-2.7/check_libyaml.o
    build/temp.linux-armv7l-2.7/check_libyaml.c:2:18: fatal error: yaml.h: No such file or directory
     #include <yaml.h>
                      ^
    compilation terminated.

    libyaml is not found or a compiler error: forcing --without-libyaml
    (if libyaml is installed correctly, you may need to
     specify the option --include-dirs or uncomment and
     modify the parameter include_dirs in setup.cfg)
    error: could not create '/usr/local/lib/python2.7/dist-packages/yaml': Permission denied
    Complete output from command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-build-7bLi6E/pyyaml/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-hCD07D-record/install-record.txt --single-version-externally-managed --compile:
    running install

running build

running build_py

creating build

creating build/lib.linux-armv7l-2.7

creating build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/tokens.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/serializer.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/events.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/constructor.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/composer.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/cyaml.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/emitter.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/resolver.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/loader.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/parser.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/scanner.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/reader.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/error.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/dumper.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/nodes.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/__init__.py -> build/lib.linux-armv7l-2.7/yaml

copying lib/yaml/representer.py -> build/lib.linux-armv7l-2.7/yaml

running build_ext

creating build/temp.linux-armv7l-2.7

checking if libyaml is compilable

arm-linux-gnueabihf-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I/usr/include/python2.7 -c build/temp.linux-armv7l-2.7/check_libyaml.c -o build/temp.linux-armv7l-2.7/check_libyaml.o

build/temp.linux-armv7l-2.7/check_libyaml.c:2:18: fatal error: yaml.h: No such file or directory

 #include <yaml.h>

                  ^

compilation terminated.



libyaml is not found or a compiler error: forcing --without-libyaml

(if libyaml is installed correctly, you may need to

 specify the option --include-dirs or uncomment and

 modify the parameter include_dirs in setup.cfg)

running install_lib

creating /usr/local/lib/python2.7/dist-packages/yaml

error: could not create '/usr/local/lib/python2.7/dist-packages/yaml': Permission denied

----------------------------------------
Cleaning up...
Command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-build-7bLi6E/pyyaml/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-hCD07D-record/install-record.txt --single-version-externally-managed --compile failed with error code 1 in /tmp/pip-build-7bLi6E/pyyaml
Storing debug log for failure in /home/pi/.pip/pip.log

```

--- a/homeassistant/util/yaml.py
+++ b/homeassistant/util/yaml.py
@@ -43,6 +43,9 @@ def load_yaml(fname: str) -> Union[List, Dict]:
     except yaml.YAMLError as exc:
         _LOGGER.error(exc)
         raise HomeAssistantError(exc)
+    catch UnicodeDecodeError as exc:
+        _LOGGER.error('Unable to read file %s: %s', fname, exc)
+        raise HomeAssistantError
 
 
 def clear_secret_cache() -> None:

--- a/homeassistant/util/yaml.py
+++ b/homeassistant/util/yaml.py
@@ -45,7 +45,7 @@ def load_yaml(fname: str) -> Union[List, Dict]:
         raise HomeAssistantError(exc)
     catch UnicodeDecodeError as exc:
         _LOGGER.error('Unable to read file %s: %s', fname, exc)
-        raise HomeAssistantError
+        raise HomeAssistantError(exc)
 
 
 def clear_secret_cache() -> None:

--- a/homeassistant/util/yaml.py
+++ b/homeassistant/util/yaml.py
@@ -43,7 +43,7 @@ def load_yaml(fname: str) -> Union[List, Dict]:
     except yaml.YAMLError as exc:
         _LOGGER.error(exc)
         raise HomeAssistantError(exc)
-    catch UnicodeDecodeError as exc:
+    except UnicodeDecodeError as exc:
         _LOGGER.error('Unable to read file %s: %s', fname, exc)
         raise HomeAssistantError(exc)
 

--- a/tests/util/test_yaml.py
+++ b/tests/util/test_yaml.py
@@ -240,6 +240,7 @@ class TestYaml(unittest.TestCase):
 
     @patch('homeassistant.util.yaml.open')
     def test_load_yaml_encoding_error(self, mock_open):
+        """Test raising a UnicodeDecodeError."""
         mock_open.side_effect = UnicodeDecodeError('', b'', 1, 0, '')
         self.assertRaises(HomeAssistantError, yaml.load_yaml, 'test')
 

--- a/tests/util/test_yaml.py
+++ b/tests/util/test_yaml.py
@@ -237,8 +237,7 @@ class TestYaml(unittest.TestCase):
                     "key4": "four"
                 }
 
-
-    @patch('homeassistant.util.yaml.open')
+    @patch('homeassistant.util.yaml.open', create=True)
     def test_load_yaml_encoding_error(self, mock_open):
         """Test raising a UnicodeDecodeError."""
         mock_open.side_effect = UnicodeDecodeError('', b'', 1, 0, '')

--- a/tests/util/test_yaml.py
+++ b/tests/util/test_yaml.py
@@ -95,7 +95,6 @@ class TestYaml(unittest.TestCase):
         mock_walk.return_value = [
             ['/tmp', ['tmp2', '.ignore', 'ignore'], ['zero.yaml']],
             ['/tmp/tmp2', [], ['one.yaml', 'two.yaml']],
-            ['/tmp/.ignore', [], []],
             ['/tmp/ignore', [], ['.ignore.yaml']]
         ]
 
@@ -136,7 +135,6 @@ class TestYaml(unittest.TestCase):
         mock_walk.return_value = [
             ['/tmp', ['tmp2', '.ignore', 'ignore'], ['first.yaml']],
             ['/tmp/tmp2', [], ['second.yaml', 'third.yaml']],
-            ['/tmp/.ignore', [], []],
             ['/tmp/ignore', [], ['.ignore.yaml']]
         ]
 
@@ -175,7 +173,6 @@ class TestYaml(unittest.TestCase):
         mock_walk.return_value = [
             ['/tmp', ['tmp2', '.ignore', 'ignore'], ['first.yaml']],
             ['/tmp/tmp2', [], ['second.yaml', 'third.yaml']],
-            ['/tmp/.ignore', [], []],
             ['/tmp/ignore', [], ['.ignore.yaml']]
         ]
 
@@ -218,7 +215,6 @@ class TestYaml(unittest.TestCase):
         mock_walk.return_value = [
             ['/tmp', ['tmp2', '.ignore', 'ignore'], ['first.yaml']],
             ['/tmp/tmp2', [], ['second.yaml', 'third.yaml']],
-            ['/tmp/.ignore', [], []],
             ['/tmp/ignore', [], ['.ignore.yaml']]
         ]
 
@@ -240,6 +236,13 @@ class TestYaml(unittest.TestCase):
                     "key3": "three",
                     "key4": "four"
                 }
+
+
+    @patch('homeassistant.util.yaml.open')
+    def test_load_yaml_encoding_error(self, mock_open):
+        mock_open.side_effect = UnicodeDecodeError('', b'', 1, 0, '')
+        self.assertRaises(HomeAssistantError, yaml.load_yaml, 'test')
+
 
 FILES = {}
 

